Add a parallel implementation

tinou98 opened the pull request:
Added a parallel implementation so everything is faster. It's based on rayon, so it will default to the same number of parallel task as your number of CPU.
I also added a progress bar for clone and fetch

tinou98 committed:
Parallel Implementation

jpninanjohn commented on src/fetch.rs:
```diff
@@ -29,117 +32,118 @@ pub fn fetch(args: InputArgs, filter_list: Vec<Regex>) {
-    // TODO: handle all remotes
-    if remotes.iter().any(|remote| remote == Some("origin")) {
-        let remote = "origin";
-        print!("{} {} {} {:#?} -> ", "\nFetching".blue(), remote.blue(), "for".blue(), path);
```
Can you add this info that it is fetching for origin somewhere. When we add fetch for other remotes, we can show which remote we are fetching for.

tinou98 replied:
I moved the remote name up to the fetch function, so we can change it for every repo. I'm not sure of what will happen if we try to pull different origin from the same repo at the same time ? But as long as they are only one remote per repo, it should be safe.

jpninanjohn replied:
We wont pull at same time, it will fetch for one remote only but we will provide an option to fetch for a diffrent remote other than origin. At that time we need to show whihc remote we are pulling from.

tinou98 replied:
It now print the remote before the file path, and it's controlled in the parameter of the GitAction, so it's easy to add a Clap flag that control it

jpninanjohn commented on src/create.rs:
```diff
@@ -50,27 +60,27 @@ pub fn create(args: InputArgs) {
     };
 
-    clone.git_action().expect(format!("Failed to clone repo {}, ", remote_url).as_str());
+    let multi_bars = ProgressTracker::new();
```
I am not able to see a progress bar for create. In case it takes time there is no output shown. Only the final result after clone is printed.

tinou98 replied:
It might need a bit more refactor : the progress bar is made for tracking things that happen in a `GitAction`, and most of the creation step happen from outside. Either my progress system must be more flexible, either we need to put everything in a `GitAction`

jpninanjohn replied:
The wait is actually when cloning after remote creation is done. If we add a line for clone before download begins, it should be fine.

jpninanjohn commented on src/progress.rs:
```diff
@@ -0,0 +1,82 @@
+const STYLE_LOAD: &str = "{prefix:>40!.blue} {msg} {wide_bar} {percent:>3}% {eta}";
+const STYLE_DONE: &str = "{prefix:>40!.blue} {wide_msg} {elapsed_precise}";
+const STYLE_ERROR: &str = "{prefix:>40!.blue} {wide_msg:.red}";
+
```
Can we make the alignment either completely left aligned or justified?

tinou98 replied:
Left align : (use ^ for centering)
```suggestion
const STYLE_LOAD: &str = "{prefix:<40!.blue} {msg} {wide_bar} {percent:>3}% {eta}";
const STYLE_DONE: &str = "{prefix:<40!.blue} {wide_msg} {elapsed_precise}";
const STYLE_ERROR: &str = "{prefix:<40!.blue} {wide_msg:.red}"; 
```

jpninanjohn replied:
```suggestion
const STYLE_PRELOAD: &str = "{prefix:.blue} {spinner} {wide_msg:.cyan}";
const STYLE_LOAD: &str = "{prefix:.blue} {msg} {wide_bar} {percent:>3}% {eta}";
const STYLE_DONE: &str = "{prefix:.blue} {wide_msg} {elapsed_precise}";
const STYLE_ERROR: &str = "{prefix:.blue} {wide_msg:.red}";
```
I would suggest not giving a size to the prefix. The paths can be of any length and will be truncated if it doesnt fit right now. Let it always print the entire path and then the message.

jpninanjohn replied:
Can you do this change as part of ur fork as well since you will still be working on it? Or should I make the change once I merge?

jpninanjohn commented on src/clone.rs:
```diff
@@ -81,45 +91,52 @@ pub fn clone(args: InputArgs, mut clone_repos: Vec<GitRepo>) {
+            local_path: remote.local_path.into(),
+        })
+        .for_each(|clone| multi_bars.start_task(clone));
+
```
There seems to be a time period where no progress is shown before the download actually begins. Only the remote is printed. Can we show some somethign that indicates a download is about to begin??

tinou98 replied:
By default it should print Waiting for process to begin, we might print the remote message (things like
```
remote: Enumerating objects: 6915, done.
remote: Counting objects: 100% (4012/4012), done.
remote: Compressing objects: 100% (2547/2547), done.
remote: Total 2966 (delta 2589), reused 487 (delta 414)
```
), but in my experience, it was only disturbing the screen

jpninanjohn replied:
Yeah, dont print that. just a line saying cloning has begun should suffice. The clone command already has this, but its outside ur progress bar which is fine. For create, there is no feedback as to what is happening. After remote creation, can we just print that cloning will begin.

tinou98 replied:
Except for the bug in single threaded mode, it should now print `Waiting for process to begin` while waiting for a free thread, and `Process is about to begin` before the download really start

jpninanjohn commented:
Thanks for the contribution. This is really useful. I have added some comments for some minor changes. Do let me know your thoughts.

jpninanjohn commented:
Tagging @dineshba

tinou98 committed:
Move remote to the param of GitAction

jpninanjohn commented:
@tinou98 there seems to be another issue with the clone command. Previously if we try to clone the same repo to the same local path, it would fail saying the path already exists which is expected. Now it just stalls. Can you check what is happening? 
Run `gg clone -r git@github.com:thecasualcoder/gg.git -l some_dir` twice.
Expect second time to fail -> 
```
Cloning remotes passed as arguments
No remotes configured in conf file
thread 'main' panicked at 'Failed to clone repo git@github.com:thecasualcoder/gg.git, : Error { code: -4, klass: 3, message: "\'some_dir/gg\' exists and is not an empty directory" }', src/clone.rs:95:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
``` 
(error is not currently handled properly. But it should still fail)

On your fork it doesnt fail. It just gets stuck.

I also tried it for a repo that does not exist. It stalls.

jpninanjohn commented:
Also I think parallelism should be an option. If enabled use based on number of CPU's or user input count. If disabled fall back to single threaded execution. I worry that some systems might hang if all CPU's are used for gg.

tinou98 committed:
Add initial message, fix updating frequency, and add origin name in the report line

tinou98 committed:
Add control over the thread count (and ability to disable it)

tinou98 commented:
Fixed blocked issue, added control over the number of thread, setting to 1 go back to the monothreaded way. There is still an issue in monothreaded mode (`-j1`) that add latency before the status line is shown, I'm working on it

jpninanjohn commented:
It's Good. You can continue to work on it. Ill merge these now.
